src/pybind/mgr/cli_api_commands/module.py

```diff
@@ -0,0 +1,233 @@
+import functools
+import inspect
+import json
+import logging
+import time
+import threading
+from queue import Queue
+from threading import Event
+from typing import Any, Callable, Dict, List
+
+from mgr_module import MgrModule, HandleCommandResult, CLICommand, API
+
+logger = logging.getLogger()
+
+
+class CephCommander:
+    '''
+    Utility class to inspect Python functions and generate corresponding
+    CephCommand signatures (see src/mon/MonCommand.h for details)
+    '''
+    def __init__(self, func: Callable):
+        self.func = func
+        self.signature = inspect.signature(func)
+        self.params = self.signature.parameters
+
+    def to_ceph_signature(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+    def to_ceph_benchmark(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli benchmark {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+
+class ThreadedBenchmarkRunner:
+    def __init__(self, func_name: str, data: list, number_of_total_calls: int, number_of_parallel_calls: int):
+        self._func_name = func_name
+        self._data = data
+        self._number_of_total_calls = number_of_total_calls
+        self._number_of_parallel_calls = number_of_parallel_calls
+        self._threads = []
+        self._jobs: Queue = Queue()
+        self.time_accumlated = 0.0
+        self._time = []
+        self._lock = threading.Lock()
+        self._mgr = None
+
+    def set_mgr(self, mgr):
+        self._mgr = mgr
+
+    def benchmark(self):
+        self.start(self._func_name, self._data)
```

Comment:
BTW if the data is already stored in an object variable, there's no need to pass it via argument (the `start()` method can simply pick those from `self.func_name` etc.